### PR TITLE
Spatial_Engine: Add support for Cellular sections

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### NOTE: Depends on 

https://github.com/BHoM/BHoM/pull/1486   

### Issues addressed by this PR

Adds initial methods supporting CellularSection classes added in oM.

### Test files

https://burohappold.sharepoint.com/:f:/s/BHoM/Egum0nEYkI9LkPQ6ETB3SDYBd5H2x0SZhXBS9DVdz7K-Kg?e=9KCZ4i

### CODE REVIEW

File: Spatial_Engine/Query/DistributedOpeningCurves.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace BH.Engine.Spatial
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Distributes a series of openings along a centreline. Method will fit in as many openings along the curve as it can, starting from the start of the curve.")]
```
Comment (peterjamesnugent) [inline Spatial_Engine/Query/DistributedOpeningCurves.cs:43]:
```suggestion
        [Description("Distributes a series of openings along a centreline from the start point.")]
```

File: Spatial_Engine/Query/DistributedOpeningCurves.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        [Input("opening", "CellularOpening do ditribute along curve.")]
```
Comment (peterjamesnugent) [inline Spatial_Engine/Query/DistributedOpeningCurves.cs:44]:
```suggestion
        [Input("opening", "CellularOpening to distribute along curve.")]
```

File: Spatial_Engine/Query/DistributedOpeningCurves.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        [Input("centreline", "Centreline curve to distribute the openings along.")]
+        [Input("normal", "Normal direction of the element the openings belong to. The openings will be in a plane spanned by the tangent of the centreline and the normal vector.")]
```
Comment (peterjamesnugent) [inline Spatial_Engine/Query/DistributedOpeningCurves.cs:46]:
```suggestion
        [Input("normal", "Normal direction of the element the openings belong to.")]
```
The second half seems really confusing - we're basically saying it is the normal of the centreline?

File: Spatial_Engine/Query/OpeningArea.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+namespace BH.Engine.Spatial
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Returns the area of a single opening.")]
+        [Input("opening", "The cellular opening to fetch the area from.")]
```
Comment (peterjamesnugent) [inline Spatial_Engine/Query/OpeningArea.cs:41]:
```suggestion
        [Input("opening", "The cellular opening to calculate the area of.")]
```

File: Spatial_Engine/Query/OpeningArea.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                return double.NaN;
+            }
+            double r = opening.Diameter / 2;
+            return r * r * Math.PI;
+        }
+
+        /***************************************************/
+
+        [Description("Returns the area of a single opening.")]
+        [Input("opening", "The cellular opening to fetch the area from.")]
```
Comment (peterjamesnugent) [inline Spatial_Engine/Query/OpeningArea.cs:57]:
```suggestion
        [Input("opening", "The cellular opening to calculate the area of.")]
```

File: Spatial_Engine/Query/OpeningCurve.cs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+namespace BH.Engine.Spatial
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Returns the curve of a single opening in the XY plane.")]
+        [Input("opening", "The cellular opening to fetch the opening curve from.")]
```
Comment (peterjamesnugent) [inline Spatial_Engine/Query/OpeningCurve.cs:43]:
```suggestion
        [Input("opening", "The cellular opening to get the opening curve from.")]
```

File: Spatial_Engine/Query/OpeningArea.cs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            double w = opening.SinusoidalLength * 2 + opening.WidthWebPost;
+            return opening.Height * (opening.SinusoidalLength + opening.WidthWebPost);
+        }
+
+        /***************************************************/
+        /**** Public Methods - Interface                ****/
+        /***************************************************/
+
+        [Description("Returns the area of a single opening.")]
+        [Input("opening", "The cellular opening to fetch the area from.")]
```
Comment (peterjamesnugent) [inline Spatial_Engine/Query/OpeningArea.cs:90]:
```suggestion
        [Input("opening", "The cellular opening to calculate the area of.")]
```

File: Spatial_Engine/Query/OpeningArea.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+                Base.Compute.RecordError("Unable to query the area from a null opening.");
+                return double.NaN;
+            }
+            return opening.Height * (opening.Width + opening.WidthWebPost) / 2 + opening.SpacerHeight * opening.Width;
+        }
+
+        /***************************************************/
+
+        [Description("Returns the area of a single opening.")]
+        [Input("opening", "The cellular opening to fetch the area from.")]
```
Comment (peterjamesnugent) [inline Spatial_Engine/Query/OpeningArea.cs:72]:
```suggestion
        [Input("opening", "The cellular opening to calculate the area of.")]
```

File: Spatial_Engine/Query/OpeningCurve.cs [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+                return null;
+            }
+            double r = opening.Diameter / 2;
+            return new Circle { Centre = Point.Origin, Normal = Vector.ZAxis, Radius = r };
+        }
+
+        /***************************************************/
+
+        [Description("Returns the curve of a single opening in the XY plane.")]
+        [Input("opening", "The cellular opening to fetch the opening curve from.")]
```
Comment (peterjamesnugent) [inline Spatial_Engine/Query/OpeningCurve.cs:59]:
```suggestion
        [Input("opening", "The cellular opening to get the opening curve from.")]
```

File: Spatial_Engine/Query/OpeningCurve.cs [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+                };
+            }
+
+            return new Polyline { ControlPoints = ctrlPts };
+        }
+
+        /***************************************************/
+
+        [Description("Returns the curve of a single opening in the XY plane.")]
+        [Input("opening", "The cellular opening to fetch the opening curve from.")]
```
Comment (peterjamesnugent) [inline Spatial_Engine/Query/OpeningCurve.cs:106]:
```suggestion
        [Input("opening", "The cellular opening to get the opening curve from.")]
```

File: Spatial_Engine/Query/OpeningCurve.cs [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+            {
+                Curves = new List<ICurve>
+                {
+                    s1,
+                    s2,
+                    new Line{ Start = s2.EndPoint(), End = s3.StartPoint() },
+                    s3,
+                    s4,
+                    new Line{ Start = s4.EndPoint(), End = s1.StartPoint() }
+                }
```
Comment (peterjamesnugent) [inline Spatial_Engine/Query/OpeningCurve.cs:133]: Happy that the height is correct - but the length/width is not: (this is for a width of 0.6m).
image

File: Structure_Engine/Compute/IntegrateSection.cs [hunk 12]
```diff
@@ -198,16 +199,24 @@ public static partial class Compute
             results["Welz"] = results["Iz"] / Math.Max(results["Vy"], results["Vpy"]);
             results["Wely"] = results["Iy"] / Math.Max(results["Vz"], results["Vpz"]);
 
-            if (discontinius)
+            if (computeShearArea)
             {
-                Engine.Base.Compute.RecordWarning("Asy and Asz are not calculated for discontinuous sections. They have ben set to 0");
-                results["Asy"] = 0;
-                results["Asz"] = 0;
+                if (discontinius)
```
Comment (peterjamesnugent) [inline Structure_Engine/Compute/IntegrateSection.cs:204]:
```suggestion
                if (discontinuous)
```

File: Structure_Engine/Compute/IntegrateSection.cs [hunk 13]
```diff
@@ -198,16 +199,24 @@ public static partial class Compute
...
+                {
+                    Engine.Base.Compute.RecordWarning("Asy and Asz are not calculated for discontinuous sections. They have ben set to 0");
```
Comment (peterjamesnugent) [inline Structure_Engine/Compute/IntegrateSection.cs:206]:
```suggestion
                    Engine.Base.Compute.RecordWarning("Asy and Asz are not calculated for discontinuous sections. They have ben set to 0");
```

File: Structure_Engine/Create/SectionProperties/CellularSection/CellularSectionFromBaseSection.cs [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+
+            if (baseProfile == null)
+            {
+                Engine.Base.Compute.RecordError($"Can only create Cellular beams from SteelSections with profiles of type {nameof(ISectionProfile)}");
+                return null;
+            }
+
+            //Bring name over from baseSection to baseProfile
+            if (string.IsNullOrEmpty(baseProfile.Name) && !string.IsNullOrEmpty(baseSection.Name))
+                baseProfile.Name = baseSection.Name;
```
Comment (peterjamesnugent) [inline Structure_Engine/Create/SectionProperties/CellularSection/CellularSectionFromBaseSection.cs:63]: Do we want to bring over `tags` and `Fragments`? I know it's not commonly used on `Profiles` but may just be worth porting over the `Base` properties for completeness?

File: Structure_Engine/Create/SectionProperties/CellularSection/CellularSectionFromBaseSection.cs [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
```
Comment (peterjamesnugent) [inline Structure_Engine/Create/SectionProperties/CellularSection/CellularSectionFromBaseSection.cs:100]: As above.

File: Structure_Engine/Query/Mass.cs [hunk 16]
```diff
@@ -34,12 +36,17 @@ public static partial class Query
         /**** Public Methods                            ****/
         /***************************************************/
 
-        [Description("Calculates the mass of a Bar as its length times mass per metre. The mass per metre is generally calculated as section area multiplied by the density. No offsets or similar ar taken into acount.")]
+        [Description("Calculates the mass of a Bar as its solid volume (generally take as length times section area) times density(ies) of its material. No offsets or similar ar taken into acount.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/Mass.cs:39]:
```suggestion
        [Description("Calculates the mass of a Bar as its solid volume (generally taken as length times section area) times density(ies) of its material. No offsets or similar are taken into account.")]
```

File: Structure_Engine/Query/VolumePerLength.cs [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+            if(section.IsNull())
+                return 0.0;
+
+            double solidArea = section.SolidProfile.Area();
+
+            //The reduction in volume per metre of the section
+            double openingReduction = section.Opening.IOpeningArea() / section.Opening.Spacing * section.SolidProfile.WebThickness;
+
+            if (openingReduction != 0)
+                Base.Compute.RecordNote($"The volume per length for cellular sections is a slight approximation. To get a more accurate volume for a particular bar with the section use the {nameof(Query.SolidVolume)} method instead.");
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/VolumePerLength.cs:56]:
```suggestion
                Base.Compute.RecordNote($"The volume per length for cellular sections is an approximation. To get a more accurate volume for a particular bar with the section use the {nameof(Query.SolidVolume)} method instead.");
```

**Review by peterjamesnugent — CHANGES_REQUESTED** [review]
Tested with script, checked dimensions of openings (Sinusoidal needs a check for length) and changes implemented to put `spacing` in the `Create` methods instead of `widthWebPost`.

File: Spatial_Engine/Create/ShapeProfiles/CellularProfile/CircularOpening.cs [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        public static CircularOpening CircularOpening(double diameter, double widthWebPost)
+        {
+
+            if (widthWebPost > diameter)
+            {
+                Engine.Base.Compute.RecordError($"The {nameof(diameter)} needs to be larger than {nameof(widthWebPost)}. Unable to create circular opening.");
+                return null;
+            }
+
+            double spacing = diameter + widthWebPost;
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Spatial_Engine/Create/ShapeProfiles/CellularProfile/CircularOpening.cs:52]: Yep, agreed, will change to have spacing as input for all creates instead.

File: Spatial_Engine/Query/OpeningCurve.cs (hunk 11 above)
Comment (IsakNaslundBh) [reply Spatial_Engine/Query/OpeningCurve.cs:133]: Might need to tweak descriptions/inputs here. The length refers to the length of the sinusoidal part, not the full length.

Happy to chat through how we can make this clearer.

File: Structure_Engine/Create/SectionProperties/CellularSection/CellularSectionFromBaseSection.cs (hunk 14 above)
Comment (IsakNaslundBh) [reply Structure_Engine/Create/SectionProperties/CellularSection/CellularSectionFromBaseSection.cs:63]: Not sure... But, yeah, could probably do. Will be empty for 99% of the cases, but might as well I guess.

File: Spatial_Engine/Query/OpeningCurve.cs (hunk 11 above)
Comment (peterjamesnugent) [reply Spatial_Engine/Query/OpeningCurve.cs:133]: Yeah - I guess you really want to be able to set the total width.

Comment (IsakNaslundBh) [reply Spatial_Engine/Query/OpeningCurve.cs:133]: Updated the create method do do this now

File: Structure_Engine/Compute/IntegrateSection.cs (hunk 13 above)
Comment (IsakNaslundBh) [reply Structure_Engine/Compute/IntegrateSection.cs:206]: suggested change is empty?

File: Spatial_Engine/Create/ShapeProfiles/CellularProfile/HexagonalOpening.cs [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace BH.Engine.Spatial
+{
+    public static partial class Create
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Creates a HexagonalOpening to be applied to a cellular/castelated beam.")]
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Spatial_Engine/Create/ShapeProfiles/CellularProfile/HexagonalOpening.cs]: Suggested change is showing the same as is there?

**Review by peterjamesnugent — APPROVED** [review]
Happy that further commits have addressed changes I requested.

**Review by peterjamesnugent — APPROVED** [review]
Reapprove because of unit tests.

**Comment by IsakNaslundBh** [comment]
@BHoMBot check null-handling
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am requesting neutral checks on: versioning, ready-to-merge

**Pull request merged**